### DISCUSSION TITLE
Documentation

### DISCUSSION THREAD

**Discussion by sevki** [Harvey-OS/harvey#1074] [Ideas]
Hey Folks,

IMHO our documentation is really lacking. 

Most of the docs are outdated and the code ya'll write isn't as legible as you'd like to believe :joy: 

I really miss docs experience whenever I'm writing go or rust code and it's making me want to not touch harvey code all that much. 

So let's do something about it!!!!

What my current thinking is patch `build.go` to generate compilation DB.
https://github.com/Harvey-OS/go/pull/35

this generates a `compile_commands.json` under `$HARVEY`

Once we merge that in we can modify the CI to pull the `harvey-os.github.io` repo

```
git clone git@github.com:Harvey-OS/harvey-os.github.io.git
```
after which we can generate markdown files for C code using clang-doc
```
clang-doc --output=harvey-os.github.io/docs/ --format=md -p=. --project-name=harvey --repository=https://github.com/Harvey-OS/harvey/blob/main/ ./sys/src/
```
than commit and push the docs to the repo so every time we update our code the website and docs reflect the changes. 

What do ya'll think?

**Comment by orangecms** [comment]
Absolutely, didn't even know about `clang-doc` - that's fantastic! \o/

**Comment by sevki** [comment]
Ha. Neither did I! 

I was just googling libtooling doc generator assuming someone would have written one. I was right!

Since the comment style is doxygen and none of our code (except for some apica code) is commented in that style the MD files it generates right now doesn't seem all that useful https://github.com/Harvey-OS/harvey-os.github.io/compare/master...sevki:master

but I think it's far better than no documentation at all.

**Comment by gmacd** [comment]
Looks interesting. Would this only be used by the ci?

**Reply by sevki** [reply]
it wouldn't be exclusive to the CI. no reason why you wouldn't be able to generate docs manually and push them

but practically speaking I don't expect we'll start getting vigilant about docs overnight so IMHO CI should be in charge of keeping our docs up to date

**Comment by gmacd** [comment]
Looks worthwhile. Slightly unrelated, but it would be great to get the man pages up on the website too, maybe convert them to md 😀

**Reply by sevki** [reply]
TOTALLLYYYY!!!  I don't think it's unrelated at all

I have a MD to troff converter [here](https://pkg.go.dev/sevki.org/troff) which is compatible with plan9 troff macros so after we have the MD files we can convert them back to  tr than generate the pdf and man pages with the 9 tooling

**Comment by fhs** [comment]
Agree on the man page. Some of the kernel internals are documented in man pages in [Section 10 in Inferno](http://man.cat-v.org/inferno/10/) and [Section 9 in 9front](http://man.cat-v.org/9front/9/). Maybe we can copy and update those.

**Reply by sevki** [reply]
YES PLEASE!!
